lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js

```diff
@@ -181,6 +181,189 @@
         }
     });
 
+    platform.registerChakraLibraryFunction("MergeSort", function(array, length, compareFn) {
+        const buffer = [];
+        buffer.__proto__ = null;
+
+        let i = 1, j = 0, k = 0, bucketSize = 2, lastSize = 1, position = 0;
+
+        // handle pairs first
+        const doubleLength = length * 2;
+
+        while (bucketSize < doubleLength) {
+            while (position < length) {
+                const left = position;
+                const mid = left + lastSize;
+
+                // perform a merge but only if it's necessary
+                if (mid < length && compareFn(array[mid], array[mid - 1]) < 0) {
+                    let right = position + bucketSize;
+                    right = right < length ? right : length;
+                    i = left, j = 0, k = left;
+
+                    while (i < mid) {
+                        buffer[j++] = array[i++];
+                    }
+    
+                    const bufferLen = j;
+                    j = 1;
+                    i = mid + 1;
+                    let rightElement = array[mid];
+                    let leftElement = buffer[0];
+
+                    for (;;)
+                    {
+                        if (compareFn(rightElement, leftElement) < 0) {
+                            array[k++] = rightElement;
+                            if (i < right) {
+                                rightElement = array[i++];
+                            } else {
+                                array[k++] = leftElement;
+                                break;
+                            }
+                        } else {
+                            array[k++] = leftElement;
+                            if (j < bufferLen) {
+                                leftElement = buffer[j++];
+                            } else {
+                                break;
+                            }
+                        }
+                    }
+
+                    while (j < bufferLen) {
+                        array[k++] = buffer[j++];
+                    }
+                }  
+                position += bucketSize;
+            }
+            position = 0;
+            lastSize = bucketSize;
+            bucketSize *= 2;
+        }
+    });
+
+    platform.registerChakraLibraryFunction("DefaultSortCompare", function(left, right) {
+        // as the sort only ever checks for < 0 on the return value of compare functions
+        // only have to handle this case
+        if (left.string < right.string) {
+            return -1;
+        }
+        return 0;
+    });
+
+    platform.registerChakraLibraryFunction("CreateCompareArray", function(array, length) {
+        const compArray = [];
+        compArray.__proto__ = null;
+        let i = 0;
+        let value;
+        while (i < length) {
+            value = array[i];
+            compArray[i++] = {
```

Comment:
Would we gain any noticeable improvement on arrays of strings using default comparison if we tried to detect that case and avoided creating these extra objects? It would mean an extra walk through the array up-front doing `typeof` checks, but I imagine that would still be faster than allocating all of these temporary objects.

Reply:
But would typeof be detectable e.g. via a proxy - per spec array.prototype.sort doesn’t use typeof. I suppose I could make a native method that does this?

Reply:
I don’t think `typeof` can trigger any side effects but I’d have to double-check the specification to be sure.